cosmo_tester/framework/util.py

```diff
@@ -414,44 +402,113 @@ def generate_ca_cert(ca_cert_path, ca_key_path):
-            )
-        )
-    logger.info('Execution complete')
+    current_time = time.time()
+    # Timeout after ~5 minutes
+    timeout_time = current_time + timeout
+    password_updated = False
+    output_events(manager, execution, logger, to_time=current_time)
+
+    with set_client_tenant(manager, tenant):
+        while True:
+            try:
+                execution = manager.client.executions.get(execution['id'])
+            except UserUnauthorizedError:
+                if new_password and not password_updated:
+                    # This will happen on a restore with modified users
+                    change_rest_client_password(manager, new_password)
+                    password_updated = True
+                else:
+                    # We either shouldn't change the password, or did already.
+                    raise
+
+            prev_time = current_time
+            current_time = time.time()
+
+            output_events(manager, execution, logger, prev_time, current_time)
+
+            if time.time() >= timeout_time:
+                raise ExecutionTimeout(
+                    'Execution timed out in state: {status}'.format(
+                        status=execution.status,
+                    )
+                )
+
+            if execution.status in execution.END_STATES:
+                # Give time for any last second events
+                time.sleep(2)
+                output_events(manager, execution, logger,
+                              current_time)
+
+                if execution.status != execution.TERMINATED:
+                    logger.warning('Execution failed')
+                    raise ExecutionFailed(
+                        '{status}: {error}'.format(
+                            status=execution.status,
+                            error=execution['error'],
+                        )
+                    )
+
+                logger.info('Execution completed in state {status}'.format(
+                        status=execution.status,
+                    )
+                )
+                break
+
     return execution
+
+
+def output_events(manager, execution, logger, from_time=None, to_time=None):
+    events = manager.client.events.list(
+        execution_id=execution.id,
+        _size=1000,
+        include_logs=True,
+        sort='reported_timestamp',
+        from_datetime=convert_epoch_to_time_string(from_time),
+        to_datetime=convert_epoch_to_time_string(to_time),
+    )
+    log_methods = {
+        'debug': logger.debug,
+        'info': logger.info,
+        'warn': logger.warn,
+        'warning': logger.warn,
+        'error': logger.error,
+    }
+    for event in events:
+        if event.get('type') == 'cloudify_event':
+            level = 'info'
+        else:
+            level = event.get('level')
+
+        if level not in log_methods:
+            logger.warn('Event level {} was unknown.'.format(level))
+            logger.warn('Event was: {}'.format(event))
```

Comment:
```suggestion
            logger.warn('Unknown event level: %s', level)
            logger.warn('Event was: %s', event)
```
the usual logger/format, but this also rewords the first message, because messages scan better if the thing you're looking for isn't between words (at the beginning would be even better, but at the end is also OK)

Reply:
%s already updated in a later commit. I'll update the log structure though, yes, good point.